### PR TITLE
[JENKINS-34468] CLI command offline-node extracted from core to CLI

### PR DESCRIPTION
**Pull request by pjanouse**
offline-node extended to accept multiple node names
offline-node covered by test-cases

### CODE REVIEW

File: core/src/main/java/hudson/cli/OfflineNodeCommand.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Override
+    protected int run() throws Exception {
+        boolean errorOccurred = false;
+        final Jenkins jenkins = Jenkins.getActiveInstance();
+        final HashSet<String> hs = new HashSet<String>(nodes);
+        List<String> names = null;
+
+        for (String node_s : hs) {
+            Computer computer = null;
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/OfflineNodeCommand.java]: From what I see toy do not use <code>computer</code> outside of try-catch

File: core/src/main/java/hudson/cli/OfflineNodeCommand.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+            try {
+                computer = jenkins.getComputer(node_s);
+                if (computer == null) {
+                    if (names == null) {
+                        names = new ArrayList<String>();
+                        for (Computer c : jenkins.getComputers()) {
+                            if (!c.getName().isEmpty()) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/OfflineNodeCommand.java]: So it's a check for the master node

File: core/src/main/java/hudson/cli/OfflineNodeCommand.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        final Jenkins jenkins = Jenkins.getActiveInstance();
+        final HashSet<String> hs = new HashSet<String>(nodes);
+        List<String> names = null;
+
+        for (String node_s : hs) {
+            Computer computer = null;
+
+            try {
+                computer = jenkins.getComputer(node_s);
+                if (computer == null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/OfflineNodeCommand.java]: I would move "findNearest" implementation to a standalone Util method

**Comment by oleg-nenashev** [comment]
Test failure is really related to the change. Needs fix

File: test/src/test/java/hudson/cli/OfflineNodeCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+                .invokeWithArgs("aNode1", "aNode2", "aNode3");
+        assertThat(result, succeededSilently());
+        assertThat(slave1.toComputer().isOffline(), equalTo(true));
+        assertThat(slave1.toComputer().getOfflineCause() instanceof OfflineCause.ByCLI, equalTo(true));
+        assertThat(((OfflineCause.ByCLI) slave1.toComputer().getOfflineCause()).message, equalTo(null));
+        assertThat(slave2.toComputer().isOffline(), equalTo(true));
+        assertThat(slave2.toComputer().getOfflineCause() instanceof OfflineCause.ByCLI, equalTo(true));
+        assertThat(((OfflineCause.ByCLI) slave2.toComputer().getOfflineCause()).message, equalTo(null));
+        assertThat(slave3.toComputer().isOffline(), equalTo(true));
+        assertThat(slave3.toComputer().getOfflineCause() instanceof OfflineCause.ByCLI, equalTo(true));
```
Comment (olivergondza) [inline test/src/test/java/hudson/cli/OfflineNodeCommandTest.java]: Nit: `Matchers#instanceOf` provides better reporting.

**Comment by olivergondza** [comment]
:+1:

File: test/src/test/java/hudson/cli/OfflineNodeCommandTest.java (hunk 4 above)
Comment (pjanouse) [inline test/src/test/java/hudson/cli/OfflineNodeCommandTest.java]: :+1: You are absolutely right, I've been using that Matcher for awhile, this has to be some old code, fixing...

**Comment by oleg-nenashev** [comment]
👍

**Pull request merged**